File: packages/daemon/src/resolve.ts
import { AtError } from "@athing/sdk";
import { spawnSync } from "node:child_process";

function loginShellWhich(binary: string): string | null {
  const shell = process.env["SHELL"] ?? "/bin/sh";
  const result = spawnSync(shell, ["-lc", `which ${binary}`], {
    encoding: "utf8",
    timeout: 5000,
  });
  if (result.status === 0) {
    const resolved = result.stdout.trim();
    if (resolved) return resolved;
  }
  return null;
}

/**
 * Generic command resolution for the terminal backend. An absolute path is used
 * as given; a bare name is resolved via the login-shell PATH; when no command is
 * supplied the user's login shell is used. The daemon carries no application
 * default command, install location, or version gate.
 */
export function resolveCommand(command?: string): string {
  if (!command) return process.env["SHELL"] ?? "/bin/sh";
  if (command.startsWith("/")) return command;

  const fromShell = loginShellWhich(command);
  if (fromShell) return fromShell;

  throw new AtError(
    "BinaryNotFound",
    `Cannot resolve '${command}'. Provide an absolute path or ensure it is on PATH.`,
  );
}

File: packages/daemon-rs/src/resolve.rs
//! Generic command resolution: absolute path as-is; bare name via login-shell PATH;
//! no command defaults to `$SHELL`. Unresolvable named command → `BinaryNotFound`.

use std::process::Command;

#[derive(Debug)]
pub struct BinaryNotFound(pub String);

impl std::fmt::Display for BinaryNotFound {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.0)
    }
}
impl std::error::Error for BinaryNotFound {}

fn login_shell() -> String {
    std::env::var("SHELL").unwrap_or_else(|_| "/bin/sh".to_string())
}

fn login_shell_which(binary: &str) -> Option<String> {
    let shell = login_shell();
    let out = Command::new(&shell)
        .args(["-lc", &format!("which {binary}")])
        .output()
        .ok()?;
    if out.status.success() {
        let resolved = String::from_utf8_lossy(&out.stdout).trim().to_string();
        if !resolved.is_empty() {
            return Some(resolved);
        }
    }
    None
}

pub fn resolve_command(command: Option<&str>) -> Result<String, BinaryNotFound> {
    let Some(command) = command.filter(|c| !c.is_empty()) else {
        return Ok(login_shell());
    };
    if command.starts_with('/') {
        return Ok(command.to_string());
    }
    if let Some(found) = login_shell_which(command) {
        return Ok(found);
    }
    Err(BinaryNotFound(format!(
        "Cannot resolve '{command}'. Provide an absolute path or ensure it is on PATH."
    )))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn absolute_path_used_directly() {
        assert_eq!(
            resolve_command(Some("/usr/bin/env")).unwrap(),
            "/usr/bin/env"
        );
    }

    #[test]
    fn no_command_defaults_to_shell() {
        let resolved = resolve_command(None).unwrap();
        assert!(resolved.starts_with('/'));
    }

    #[test]
    fn unresolvable_errors() {
        assert!(resolve_command(Some("definitely-not-a-real-binary-xyz")).is_err());
    }
}
